### ISSUE TITLE
MAS Build confusion

### ISSUE THREAD

**Issue by gtamas** [electron-userland/electron-builder#4640]
* **Version**: 21.2.0

* **Target**: MacOS

I'm trying to submit an app to the Apple Store.

My understanding is that this requires a MAS build, which has to be signed with the following certificates:

3rd Party Mac Developer Installer
3rd Party Mac Developer Application

Is this all I need or do I  need a Developer ID Application certificate as well?

If I sign like this, the process appears to be successful (it complains about missing Developer ID Certificate, but it signs with the other two) and the app also works.

However, when I upload to Apple's notary service for checking, I get a bunch of errors. Basically it says the app is not signed, the hardenedRuntime is not set etc, even though I have this in config:

```
  mas: {
                    hardenedRuntime: true,
                    gatekeeperAssess: true,
                    entitlements: "build_files/entitlements.mac.plist",
                    entitlementsInherit: "build_files/entitlements.mac.plist"
                },
```

According to docs, I don't need any more certificates,  the 3rd party ones should be enough for MAS. What am I doing wrong??

**Comment by gtamas** [comment]
I'm on MacOS 10.14.6

**Comment by kzimny** [comment]
maybe this question would help? https://github.com/electron-userland/electron-builder/issues/4616

**Comment by gtamas** [comment]
@kzimny 
Thanks, I tried moving all of the options from "mas" to "mac" but it didn't help.

Actually stuff like  codesign -dv --verbose=4 dist_electron/mas/my.app shows that the app is indeed signed, yet the notary service still cannot find the signature..

Also, offline notary check fails:

spctl -v -a dist_electron/mas/my.app


Don't really know what to try. Any ideas? :)

**Comment by steve981cr** [comment]
@gtamas, I just posted a blog on the MAS build process. It will successfully build a Hello World Electron app for the Mac App Store. So that way you can tell if you have all the basics set up right. My app is still not working though because of a native dependency my app uses (Sqlite). 
https://www.techandstartup.com/tutorials/release-electron-app-on-mac-app-store

**Comment by gtamas** [comment]
@steve981cr 
Thanks, nice article! 

One question: 

I'm using electron-builder and electron-notarize. Once the build complete, is there any RELIABLE way to test the package to make sure it's properly signed AND it's truly ready to be submitted to App store? I don't want Apple to reject it, I would rather test everything offline before I send anything to Apple.

I also have a native dependency (GRPC). Please let me know if you find a solution for your native dep problem.

**Comment by steve981cr** [comment]
When you say you are using electron-notarize I assume you aren't using that for the Mac App Store build. Apps on the MAS don't need to be Notarized. 
For testing the app one option is to use a different Mac, one that doesn't have your developer credentials on it. Make sure to install the .pkg version of it, not the stand alone app.
Another option is to install a Virtual Machine on your mac (e.g., Oracle's Virtual Box). In fact I plan on trying this out later today.

**Comment by gtamas** [comment]
@steve981cr 

By following your suggestions I was able to successfully sign and build the MAS package. The app runs fine on my mac, now just I have to test it on another machine.

One question remains:

My app uses only one native dependency directly (grpc), but this command prints several more:

`find node_modules -type f -name "*.node" 2>/dev/null | grep -v "obj\.target"`

This prints a whole bunch of other node extensions besides grpc, such as different versions of node-sass, fsevents etc. Am I supposed to sign those as well? According to the official electron guide, I am supposed to sign all the native dependencies my app uses..

https://www.electronjs.org/docs/tutorial/mac-app-store-submission-guide

Well, I tried to list all of them under the "binaries" MAS config option. 

After rebuild, the app works fine, but strangely if I omit the "binaries" key, the build is still OK..
You said your app doesn't work because of the sqlite dependency. How come my app works even I don't sign any of the native dependencies?

Thanks again for the great article and the help!
